src/poller_base.cpp

```diff
@@ -67,9 +71,6 @@ void zmq::poller_base_t::cancel_timer (i_poll_events *sink_, int id_)
             _timers.erase (it);
             return;
         }
-
-    //  Timer not found.
```

Comment:
It's not that nice that this consistency check is removed.

Reply:
This code can now be reached because in `execute_timers ()` timer is deleted before `timer_event ()` is executed and `timer_event ()` might call `cancel_timer ()` on already deleted one.

It's very difficult for me to trace through all possible calls to be 100% sure but the issue itself indicates that this might happen.

Reply:
Yes, I understand why it can't be kept given the suggested changes in `execute_timers`. I just wanted to reinforce this side effect. As I said, I can imagine that other alternatives might be even worse, if it's not possible to eliminate this reentrant call altogether. And I understand that the situation leading to this is hard to analyze.

Maybe leave a comment here that "usually" this condition can be asserted except for the reentrant case? If we understand the situation better in the future, this maybe could be changed back again.

Reply:
Yes, you're right. I'll add the comment.

Reply:
I've taken another look at the code.
It seems to me now that the assert should stay.
It doesn't make sense to call `cancel_timer ()` from `timer_event ()` handler on the timer that triggered the `timer_event ()`.
Also, the proposed fix from the reported issue did not remove the assert and one of the people who tested it did not run into this assert at any moment.
So I'll put back the assert as it was.

Reply:
Reverted to as it was.

Reply:
Hm... This means then that the call to `cancel_timer ()` from `timer_event ()` for the timer on which it is being called will fail this assert.
So in that case we don't need to delete the iterator before the `timer_event ()` call nor make a copy of it's contents.
That makes the code simpler.

Reply:
But then we get to the almost the same code we started with, for which the issue was reported.
I'll put a comment then instead of the assert.

Reply:
OK, I've removed the assert again and put the comment.

Reply:
Indeed, if the assert wouldn't be hit after the changes to `execute_timers`, then I'd understand the whole problem even less. I still hope to get more insight in the future :)

Reply:
Thanks for the great comment(s)!